### ISSUE TITLE
Animated WebP Image not showing correctly using UIImageView or FLAnimatedImageView

### ISSUE THREAD

**Issue by dreampiggy** [rs/SDWebImage#1951]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios
 Platform Version        | 7.0-10.0
 SDWebImage Version      | 4.0
 Integration Method      | cocoapods
 Xcode Version           | Xcode 8.3.2
 Repro rate              | all the time (100%)
 Repro with our demo prj | YES, the demo Image-5 renders wrong.
 Demo project link       | https://github.com/dreampiggy/AnimatedWebPDemo

### Issue Description and Steps

Description:

SDWebImage 4.0 add support to [animated WebP](https://developers.google.com/speed/webp/faq#why_should_i_use_animated_webp). But when downloading and showing some of animated WebP using SDWebImage API(eg. `sd_setImageWithURL:`) to `UIImageView` or `FLAnimatedImageView` instance, the imageview can't render correctly.

Comparing to another animated WebP library ([YYImage](https://github.com/ibireme/YYImage)), it seems that the decoder for WebP to get per frame images works wrong.

Also, current implementation does not support WebP [disposal method](https://developers.google.com/speed/webp/docs/riff_container#animation),  which cause many animated WebP images look strange. And due to UIImage's API, the `duration` parameter just use the average duration for each frame, which is different from the animated WebP standard which allows set different duration per frame and also cause some images look strange.

For more detail and intuitive effect, you can go to the [demo project]( https://github.com/dreampiggy/AnimatedWebPDemo) to build and run.

Step:
1. find some animated WebP images' URLs.
2. create an `FLAnimatedImageView` or `UIImageView` instance and use `sd_setImageWithURL:` API to set image.
3. when showing animated WebP images, the images not showing correctly(looks strange).

**Comment by dreampiggy** [comment]
By reading the decoder from YYImage and SDWebImage , I found that for single frame animated WebP, the libwebp's `WebPBitstreamFeatures`.`height` and `width` is the images actual size. But for multiple frame images, this will return the current frame canvas size(but not image size). This will make the UIImages array with the wrong size. We should use`WebPDemuxGetI` to get the actual size.

Looks at this animated webp images (actual size is 100x100), but the animated images array to use for `UIImage -animatedImageWithImages:` show the wrong size.

![Xcode debug](http://7xsf4p.com1.z0.glb.clouddn.com/image/png/SDWebImage-webp-1.png)

**Comment by dreampiggy** [comment]
The YYImageCoder code reference :[YYImage](https://github.com/ibireme/YYImage/blob/6a756e93fd9fae5f5c6920d821514f3ae91b0a61/YYImage/YYImageCoder.m#L1351)

**Comment by dreampiggy** [comment]
There is an simple fix(which first create UIImage through CGImageCreate from WebPData. Then draw this UIImage to an new CGBitmapContext with actual size(not just the canvas size).

[fix screenshot](https://raw.githubusercontent.com/dreampiggy/AnimatedWebPDemo/fix_sd_animated_webp_canvas_size/ScreenShot.png)

Consider a higher performance(because create CGBitMapContext need resource), maybe we can modify the RGBA vector to do these). 

The simple fix commit [Fixed animated WebP image frame size wrong](https://github.com/dreampiggy/SDWebImage/commit/4d6b3b9cfb5927d9dfbec797dc0ac8d4e8380bbd)

**Comment by dreampiggy** [comment]
Also, I found that the animated WebP dipose method(see [here](https://developers.google.com/speed/webp/docs/riff_container#animation)) is not support, currently is assume that all frames use the background disposal.

By sharing an blend canvas while decoding multiple frame animated webp image, we could do disposal and blend correctly(not relay on that last image).

**Comment by dreampiggy** [comment]
PR is here: https://github.com/rs/SDWebImage/pull/1952

**Comment by dreampiggy** [comment]
Because `+[UIImage animatedImageWithImages:duration:]` just use the average duration for each frame, which is not the animated WebP standard. Animated WebP can has different duration per frame. Therefore, we should divide the total duration and add image item multiple times to solve this issue. The code like this:

```objective-c
NSArray<UIImage *> *)images;
NSUInteger count = images.count;
int durations[count];
int const gcd = gcdArray(count, durations);
NSMutableArray<UIImage *> *animatedImages = [NSMutableArray arrayWithCapacity:count];
[images enumerateObjectsUsingBlock:^(UIImage * _Nonnull image, NSUInteger idx, BOOL * _Nonnull stop) {
    int duration = durations[idx];
    int repeatCount = duration / gcd;
    for (int i = 0; i < repeatCount; ++i) {
        [animatedImages addObject:image];
    }
}];
```

**Comment by bpoplauschi** [comment]
Fixed by #1952
